theano/gpuarray/dnn.py

```diff
@@ -2800,6 +2800,188 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        # complaining because of the MergeOptimizer, we make this variable
+        # always compare to True.
+        out = node.outputs[0]
+        out.tag.values_eq_approx = tensor.type.values_eq_approx_always_true
+        return node
+
+    def c_code_cache_version(self):
+        return (super(GpuDnnSpatialTfDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnGridGenerator(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid generator for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dtype',)
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/spatialtf_grid.c"], "spatialtf_grid")
+        self.dtype = dtype
+
+    def make_node(self, grid_dimensions, theta, desc):
+        context_name = infer_context_name(desc, theta)
+
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert theta.dtype in ('float16', 'float32', 'float64')
+
+        # Allocate GPU memory for grid of coordinates
+        grid = GpuArrayType(dtype=self.dtype,
+                            broadcastable=(False, False, False, False,),
+                            context_name=context_name)()
+
+        return Apply(self, [grid_dimensions, theta, desc], [grid])
+
+    def L_op(self, inputs, outputs, output_grads):
+        pass
+
+
+class GpuDnnGridSampler(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid sampler for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dtype',)
+    _cop_num_inputs = 5
+    _cop_num_outputs = 1
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/spatialtf_sampler.c"], "spatialtf_sampler")
+        self.dtype = dtype
+
+    def make_node(self, img, grid, desc, alpha=None, beta=None):
+        context_name = infer_context_name(img, grid)
+
+        img = as_gpuarray_variable(img, context_name)
+        grid = as_gpuarray_variable(grid, context_name)
+
+        output = GpuArrayType(dtype=self.dtype,
+                              broadcastable=img.type.ndim * (False,),
+                              context_name=context_name)()
+
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        alpha = ensure_dt(alpha, _one, 'alpha', img.dtype)
+        beta = ensure_dt(beta, _zero, 'beta', img.dtype)
+
+        return Apply(self, [img, grid, desc, alpha, beta], [output])
+
+    def L_op(self, inputs, outputs, output_grads):
+        pass
+
+
+def dnn_spatialtf(inp, theta, downsampling_factor=1, alpha=None, beta=None, dtype=theano.config.floatX):
```

Comment:
I see you are changing the whole interface for spatial transformation by adding a `downsampling` factor. But from what I understand, an user may want to do a spatial transformation that is different for width and for height, right ? So, indeed, number of images and number of channels per image should remain the same, but there may be a new independent width and a new independent height, isn't it ?

If so, using one `downsampling` factor is a too limited strategy. I think you should better directly use params like `new_width` and `new_height`, to allow the user to explicitely specify the dimensions of the transformed images.

Or, maybe better, use params like `scale_width` and `scale_height` to allow user to explicitely specify a downsample/upsample factor for each dimension, so that:
```python
new_width = image_width * scale_width
new_height = image_height * scale_height
```
In such case (and also with current downsampling factor), I think a round strategy (probably round half up) should be applyed to get new width and new height as integers, isn't it ?